README
This is the server for Thousand Parsec.
More info at
http://www.thousandparsec.net/tp/

Requires a Unix-like environment, because it uses dynamic loading.
Tested under Linux.

REQUIRES (to build also requires the dev packages/headers)
========
pthreads (part of libc, standard)
libtprl 0.1.2 or later (replaces/uses readline) (www.thousandparsec.net/tp/) [GPL]
Boost 1.32.1 or later (www.boost.org) [Boost License]
Boost.Signals (as above)

OPTIONAL
========
libgnutls12 (common in latest distros, optional needed for secure sockets) [GPL]
Version 1.2.10 or higher.

mysql (www.mysql.com) [GPL]
MySQL is optional, but is needed for the mysql persistence backend. The client
side library is required.
 - note: see ./configure --help for options and env vars.

Avahi (avahi.org) [LGPL]
Version 0.6.0 or higher
Needed for zeroconf/mDNS support

OPTIONS (need one or the other)
=======
The configuration will try to locate these packages and will give an error if
neither is found. Both can be present and will be built.

mzscheme (www.plt-scheme.org) [LGPL]
 - note: you will probably need to pass the directory to the plt
   to configure, and it will generally be /usr/lib/plt (as in 
   debian), therefore, your configure command will be:
     ./configure --with-mzscheme=/usr/lib/plt
   You may also need to add the lib subdirectory to /etc/ld.so.conf
   and run ldconfig as root before tpserver-cpp will run.

Guile (www.gnu.org/software/guile/) [GPL]
Should be on most systems already.
Version 1.6 or higher.



SETTINGS/CONF FILE
==================
By default the settings are taken from
    /etc/tpserver-cpp/tpserver.conf
but this can be overridden by adding these command line args
    -C /dir/to/settingsfile.conf
or
    --configure /dir/to/settingsfile.conf
or
    --config_file /dir/to/settingsfile.conf
All settings can set on the console once the server is running.


SETTING UP SECURE SOCKETS
=========================
These instructions are for a self-signed certificate.

First, create private key (priv.pem):
certtool -p --outfile priv.pem

Second, create public (encrypting, tls web server?) self-signed cert 
(cert.pem):
certtool -s --load-privkey priv.pem --outfile cert.pem

In config or set on console:
x509_cert_file = cert.pem
x509_key_file = priv.pem
x509_tls = yes
tps = yes
https = yes
https_port = 8088 #or whatever


For certificates signed by a different authority, you will need the
certificate(s) of the signer and set x509_trust_file to point to that
file.


NEWS
18/04/2010 - Lee
 - Release 0.7.0 (Refactor, MTSec, BattleXML, media support)

31/03/09 - Lee
 - Release 0.6.2 (Single player support files changed, turn timers, shave)

25/03/09 - Lee
 - Release 0.6.1 (Fixes, better win32 and single player support)

03/01/09 - Lee
 - Release 0.6.0 (Risk, TaE, remote admin, MySQL persistence working again)

01/07/08 - Lee
 - Release 0.5.1 (OS X support, TP04 fix, Avahi improvements, refactoring)

01/05/08 - Lee
 - Release 0.5.0 (RFTS, TP04, very large number of changes)

01/03/07 - Lee
 - Released 0.4.1 (Non-blocking tps and https, bug fixes, Mersenne Twister (mt19937), Minisec 0.2)

31/01/07 - Lee
 - Released 0.4.0 (Non-blocking tp & http, Avahi and Metaserver support, Order parameterisation,
     Timer and Settings Callbacks)

21/11/06 - Lee
 - Released 0.3.1 (Account reg, HTTP socket, bug fixes, start of MTSec)

08/07/06 - Lee
 - Released 0.3.0 (Dynamic loading modules, resources)

19/02/06 - Lee
 - Released 0.2.1 (Minisec, config files, file/syslog logging)

12/12/05 - Lee
 - Released 0.2.0 (TP03, Minisec, Mysql backend)

14/11/05 - Lee
 - Released 0.1.2 (TP03, Minisec)

15/05/05 - Lee
 - Released 0.1.1 (Minisec with TP03, without designs)

27/01/04 - Lee
 - Release 0.1.0 (Minisec)

10/10/04 Lee Begg
 - Released 0.0.9 (minisec pre-release)

01/10/04 Lee Begg
 - Released 0.0.8

08/08/04 Lee Begg
 - Released 0.0.7

01/08/04 Lee Begg
 - Released 0.0.6 (Milestone Foxtrot)

29/07/04 Lee Begg
 - Released 0.0.6rc1 (Milestone Foxtrot)

27/05/04 Lee Begg
 - Released 0.0.5 (Milestone Echo)

02/05/04 Lee Begg
 - Released 0.0.5rc1 (Milestone Echo)

18/04/04 Lee Begg
 - Released 0.0.4 (Milestone Delta)

17/01/04 Lee Begg
 - Released 0.0.4rc2 (Milestone Delta)

07/05/03 Lee Begg
 - Released 0.0.4rc1 (Milestone Delta)

18/04/03 Lee Begg
 - Relased 0.0.3 (Milestone Charlie) 

02/03/03 Lee Begg
 - Released 0.0.2 (Milestone Bravo)

01/02/03  Lee Begg
 - Released 0.0.1 (Milestone Alpha)

Jan 03  Lee Begg
 - Initial coding
 - much of the primary server functions started
 - network, connection, frame, console and logger started
 - settings and game planning in progress


src/frame.h
#ifndef FRAME_H
#define FRAME_H
/*  TP protocol Frame class
 *
 *  Copyright (C) 2004-2005  Lee Begg and the Thousand Parsec Project
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *
 */

typedef enum {
	fv0_1 = 1,
	fv0_2 = 2,
	fv0_3 = 3,
	fv_Max,
} FrameVersion;

typedef enum {
// New Frame Codes
	ft_Invalid = -1,
	ft02_Invalid = -1,
	ft02_OK = 0,
  	ft02_Fail = 1,
	ft02_Sequence = 2,
  	ft02_Connect = 3,
  	ft02_Login = 4,
  	ft02_Object_GetById = 5,
	ft02_Object_GetByPos = 6,
  	ft02_Object = 7,
  	ft02_OrderDesc_Get = 8,
  	ft02_OrderDesc = 9,
  	ft02_Order_Get = 10,
  	ft02_Order = 11,
  	ft02_Order_Insert = 12,
  	ft02_Order_Remove = 13,
	ft02_Time_Remaining_Get = 14,
	ft02_Time_Remaining = 15,
  	ft02_Board_Get = 16,
  	ft02_Board = 17,
  	ft02_Message_Get = 18,
  	ft02_Message = 19,
  	ft02_Message_Post = 20,
	ft02_Message_Remove = 21,
	ft02_ResDesc_Get = 22,
	ft02_ResDesc = 23,
	ft02_Max = 24,
	//Version 3 additions follow
	ft03_Redirect = 24,
	ft03_Features_Get = 25,
	ft03_Features = 26,
	ft03_Ping = 27,
	ft03_ObjectIds_Get = 28,
	ft03_ObjectIds_GetByPos = 29,
	ft03_ObjectIds_GetByContainer = 30,
	ft03_ObjectIds_List = 31,
	ft03_OrderTypes_Get = 32,
	ft03_OrderTypes_List = 33,
	ft03_Order_Probe = 34,
	ft03_BoardIds_Get = 35,
	ft03_BoardIds_List = 36,
	ft03_ResType_Get = 37,
	ft03_ResType_List = 38,
	ft03_Player_Get = 39,
	ft03_Player = 40,
	//Design category and component
	ft03_Category_Get = 41,
	ft03_Category = 42,
	ft03_Category_Add = 43,
	ft03_Category_Remove = 44,
	ft03_CategoryIds_Get = 45,
	ft03_CategoryIds_List = 46,
	ft03_Design_Get = 47,
	ft03_Design = 48,
	ft03_Design_Add = 49,
	ft03_Design_Modify = 50,
	ft03_Design_Remove = 51,
	ft03_DesignIds_Get = 52,
	ft03_DesignIds_List = 53,
	ft03_Component_Get = 54,
	ft03_Component = 55,
	ft03_ComponentIds_Get = 56,
	ft03_ComponentIds_List = 57,
	ft03_Property_Get = 58,
	ft03_Property = 59,
	ft03_PropertyIds_Get = 60,
	ft03_PropertyIds_List = 61,
	ft03_Max,


} FrameType;

typedef enum {
  fec_Invalid = -1,
  fec_ProtocolError = 0,
  fec_FrameError = 1,
  fec_PermUnavailable = 2,
  fec_TempUnavailable = 3,
  fec_NonExistant = 4,
  fec_PermissionDenied = 5,
  fec_Max
} FrameErrorCode;


//class std::string;

class Frame {

	public:
		Frame();
		Frame(FrameVersion v);
		Frame(const Frame &rhs);

		~Frame();

		Frame operator=(const Frame &rhs);

		int setHeader(char *newhead);
		char *getPacket() const;
		int getHeaderLength() const;		// The length of the header section
		int getDataLength() const;		// The length of the data section
		int getLength() const;			// The total length of the packet
		
		// Data
		char *getData() const;
		bool setData(char *newdata, int dlen);

		// Type
		FrameType getType() const;
		bool setType(FrameType nt);
		
		// Sequence
		int getSequence() const;
		bool setSequence(int s);
		
		// Version
		FrameVersion getVersion() const;
		
		bool packString(const char *str);
		//bool packString(std::string str);
		bool packInt(int val);
		bool packInt64(long long val);
		bool packInt8(char val);
		bool packData(unsigned int len, char* bdata);

		// uses these functions with care
		int getUnpackOffset() const;
		bool setUnpackOffset(int newoffset);

		int unpackInt();
		char *unpackString();
		long long unpackInt64();
		char unpackInt8();
		void unpackData(unsigned int len, char* bdata);

		void createFailFrame(FrameErrorCode code, char *reason);

	private:
		FrameVersion version;
		FrameType type;
	
		// Which packet sequence does this refer to?
		int sequence;
		
		// Frame length
		int length;

		// Actual data of the frame
		char *data;

		int unpackptr;
};

#endif
